pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/FST.java

```diff
@@ -0,0 +1,438 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.utils.nativefst;
+
+import java.io.ByteArrayOutputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.BitSet;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import org.apache.pinot.segment.local.io.readerwriter.PinotDataBufferMemoryManager;
+import org.apache.pinot.segment.local.io.writer.impl.DirectMemoryManager;
+import org.apache.pinot.segment.local.utils.nativefst.builder.FSTSerializerImpl;
+
+
+/**
+ * This is a top abstract class for handling finite state automata. These
+ * automata are arc-based, a design described in Jan Daciuk's <i>Incremental
+ * Construction of Finite-State Automata and Transducers, and Their Use in the
+ * Natural Language Processing</i> (PhD thesis, Technical University of Gdansk).
+ */
+public abstract class FST implements Iterable<ByteBuffer> {
+  /**
+   * @param in The input stream.
+   * @param length Length of input to be read
+   * @return Reads remaining bytes upto length from an input stream and returns
+   * them as a byte array. Null if no data was read
+   * @throws IOException Rethrown if an I/O exception occurs.
+   */
+  protected static byte[] readRemaining(InputStream in, int length)
+      throws IOException {
+    byte[] buf = new byte[length];
+    int readLen;
+
+    readLen = in.read(buf, 0, length);
+
+    if (readLen == -1) {
+      return null;
+    }
+
+    return buf;
+  }
+
+  /**
+   * Wrapper for the main read function
+   */
+  public static FST read(InputStream stream)
+      throws IOException {
+    return read(stream, false, new DirectMemoryManager(FST.class.getName()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *DirectMemoryManager* constructor in a try-finally block. Currently, there are execution paths that do not contain closure statements, for example, when FST.read throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).